### ISSUE TITLE
Edges not available on payload when creating a custom mutation

### ISSUE THREAD

**Issue by tony-luisi** [calebmer/postgraphql#203]
When creating a custom mutation that returns a table row, the edges connection is not available. For example, if I look at the payload for a standard mutation, I get back the node and the edge:

![image](https://cloud.githubusercontent.com/assets/13082144/19910549/95cbd85e-a0f2-11e6-9a58-c6ed153a1e85.png)

However, when creating a custom mutation that returns the same thing, the edge does not get returned:

![image](https://cloud.githubusercontent.com/assets/13082144/19910564/b9413d74-a0f2-11e6-9630-a01209c67078.png)

This is useful especially when using a relay mutation (where a RANGE_ADD mutation requires an edge connection) to update the internal local state.

**Comment by calebmer** [comment]
I’m willing to add this, but with the caveat that in the future I want to deprecate all of the Relay 1 workarounds (at least workarounds that aren’t considered GraphQL best practice). However, before we add this let’s have a bit of a discussion.

Relay only needs the edge when you are adding a node to a connection (the `RANGE_ADD` mutation config). So adding an edge makes sense in the case of default insert payloads that get created automatically by PostGraphQL. However, procedures are currently opaque functions to PostGraphQL, we don’t know whether you insert, update, delete, or do all three. We only know the return type. So do we always return an edge, even for the update cases? Do we look at the function definition for an `insert` statement? Do we look at the function definition for an `insert` statement that inserts only one of the type we are returning? If we add an edge to all mutation payloads, should we add the edge field to update and delete payloads as well?

**Comment by calebmer** [comment]
Also, cc @ferdinandsalis because he’s mentioned this to me in the past :blush:

**Comment by benjie** [comment]
We could use naming conventions on mutations; e.g. if they start with `add` or `create` or whatever then assume it's effectively an insert.

**Comment by ferdinandsalis** [comment]
@benjie that could work. Using the conventions already in place it would be `create`.

Or could we also learn from introspecting the pg catalog if the mutations returns a row from a procedure (see below)? However my understanding of postgres is very limited so I will defer to @calebmer here ☺️
```sql
create function main.create_artwork_image(
  artwork_id int,
  uri text,
  priority smallint
)
returns main.image as $$
declare
  row main.image;
begin
  insert into main.image (uri) values (uri)
    returning * into row;
  insert into main.artwork_image (artwork_id, image_id, priority)
    values (artwork_id, row.id, priority);
  return row;
end;
$$ language plpgsql
strict;
```

**Comment by calebmer** [comment]
Introspection and naming conventions are both fine ideas, but I think it may be tough to communicate in practice 😣

I’m fine with just returning an edge for every procedure that returns a row object, I don’t see too much harm in doing so. Is anyone interested in implementing this? It should be fairly easy. Some relevant files:

- https://github.com/calebmer/postgraphql/blob/fc698d3da762a321971a0d8e625870847b24b609/src/graphql/schema/collection/mutations/createCreateCollectionMutationFieldEntry.ts#L53-L69
- https://github.com/calebmer/postgraphql/blob/fc698d3da762a321971a0d8e625870847b24b609/src/postgraphql/schema/procedures/createPgProcedureMutationGqlFieldEntry.ts

We have done something similar before with the top level relation fields. Relevant files for that:

- https://github.com/calebmer/postgraphql/blob/fc698d3da762a321971a0d8e625870847b24b609/src/graphql/schema/collection/mutations/createCreateCollectionMutationFieldEntry.ts#L72
- https://github.com/calebmer/postgraphql/blob/fc698d3da762a321971a0d8e625870847b24b609/src/graphql/schema/collection/createCollectionRelationTailGqlFieldEntries.ts
- https://github.com/calebmer/postgraphql/blob/fc698d3da762a321971a0d8e625870847b24b609/src/postgraphql/schema/procedures/createPgProcedureMutationGqlFieldEntry.ts#L75-L77

**Comment by ferdinandsalis** [comment]
Is anyone looking into implementing this? If not I will try my best at this …

**Comment by benjie** [comment]
@ferdinandsalis As far as I know, no-one is. Doesn't seem to be any open branches about it:

https://github.com/calebmer/postgraphql/branches/active

**Comment by calebmer** [comment]
Fixed by https://github.com/calebmer/postgraphql/pull/267. I’ll try to get a release out soon :+1:

**Comment by calebmer** [comment]
Published in `2.4.0` :tada:
